crates/ironrdp-rdpdr/src/pdu/efs.rs

```diff
@@ -0,0 +1,229 @@
+                component: Component::RDPDR_CTYP_CORE,
+                packet_id: PacketId::PAKID_CORE_CLIENTID_CONFIRM,
+            }),
+            VersionAndIdPDUKind::ServerAnnounceRequest => Err(unexpected_message_type_err!("ServerAnnounceRequest", 0)),
+        }
+    }
+
+    pub fn decode(src: &mut ReadCursor, kind: VersionAndIdPDUKind) -> PduResult<Self> {
+        Ok(Self {
+            version_major: src.read_u16(),
+            version_minor: src.read_u16(),
+            client_id: src.read_u32(),
+            kind,
+        })
+    }
+}
+
+impl PduEncode for VersionAndIdPDU {
+    fn encode(&self, dst: &mut WriteCursor<'_>) -> PduResult<()> {
+        self.header()?.encode(dst)?;
+        dst.write_u16(self.version_major);
+        dst.write_u16(self.version_minor);
+        dst.write_u32(self.client_id);
+        Ok(())
+    }
+
+    fn name(&self) -> &'static str {
+        match self.kind {
+            VersionAndIdPDUKind::ServerAnnounceRequest => "ServerAnnounceRequest",
+            VersionAndIdPDUKind::ClientAnnounceReply => "ClientAnnounceReply",
+        }
+    }
+
+    fn size(&self) -> usize {
+        // header bytes + (2 * u16) + u32 = header bytes + (2 * 2 bytes) + 4 bytes = header bytes + 8 bytes
+        self.header().unwrap().size() + 8
+    }
+}
+
+/// 2.2.2.4 Client Name Request (DR_CORE_CLIENT_NAME_REQ)
+/// https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/902497f1-3b1c-4aee-95f8-1668f9b7b7d2
+#[derive(Debug)]
+pub struct ClientNameRequest {
+    unicode_flag: ClientNameRequestUnicodeFlag,
+    computer_name: String,
+}
+
+impl ClientNameRequest {
+    pub fn new(computer_name: String) -> Self {
+        Self {
+            unicode_flag: ClientNameRequestUnicodeFlag::Unicode,
+            computer_name,
+        }
+    }
+
+    fn header(&self) -> SharedHeader {
+        SharedHeader {
+            component: Component::RDPDR_CTYP_CORE,
+            packet_id: PacketId::PAKID_CORE_CLIENT_NAME,
+        }
+    }
+}
+
+impl PduEncode for ClientNameRequest {
+    fn encode(&self, dst: &mut WriteCursor<'_>) -> PduResult<()> {
+        self.header().encode(dst)?;
+        dst.write_u32(self.unicode_flag as u32);
+        dst.write_u32(0); // // CodePage (4 bytes): it MUST be set to 0
+        dst.write_u32(encoded_str_len(&self.computer_name, self.unicode_flag.into(), true) as u32);
+        write_string_to_cursor(dst, &self.computer_name, self.unicode_flag.into(), true)
+    }
+
+    fn name(&self) -> &'static str {
+        "ClientNameRequest"
+    }
+
+    fn size(&self) -> usize {
+        // header bytes + u32 * 3 + computer_name_len = header bytes + 4 bytes * 3 + computer_name_len = header bytes + 12 + computer_name_len
+        self.header().size() + 12 + encoded_str_len(&self.computer_name, self.unicode_flag.into(), true)
+    }
+}
+
+#[repr(u32)]
+#[derive(Debug, Clone, Copy)]
+enum ClientNameRequestUnicodeFlag {
+    #[allow(dead_code)]
```

Comment:
Generally, dead code should be avoided, and PDU parsing/building code should provide the full access to all PDU variants if possible

There are a few options to avoid dead code:
- Add 2 separate public constructors for `ClientNameRequest`: `new_unicode` and `new_ascii` instead of single one
- Remove `new` in `ClientNameRequest` and make its fields public, ClientNameRequestUnicodeFlag also should be made public
- Refactor `ClientNameRequest` as described above

Reply:
`new_unicode`/`new_ascii` doesn't get rid of dead code, because then it just complains that I'm not using `new_ascii`. I was however able to get rid of the dead_code like 925c42031bc2c72d356ed14ff7c0fb7c78ed8097

Reply:
I might be missing the big picture, but remove the said dead code is also an avenue if it will never be used inside of this crate, and if it’s not intended to be part of the public API.